pydantic/_internal/_generate_schema.py

```diff
@@ -1471,7 +1471,12 @@ def _dataclass_schema(
                 dataclass = origin
 
             config = getattr(dataclass, '__pydantic_config__', None)
-            with self._config_wrapper_stack.push(config), self._types_namespace_stack.push(dataclass):
+            dataclass_bases_stack = ExitStack()
+            with self._config_wrapper_stack.push(config), dataclass_bases_stack:
+                for dataclass_base in reversed(dataclass.__mro__):
+                    if dataclasses.is_dataclass(dataclass_base):
+                        dataclass_bases_stack.enter_context(self._types_namespace_stack.push(dataclass_base))
+
```

Comment:
```suggestion
            with ExitStack() as exit_stack:
                exit_stack.enter_context(self._config_wrapper_stack.push(config))
                for dataclass_base in reversed(dataclass.__mro__):
                    if dataclasses.is_dataclass(dataclass_base):
                        exit_stack.enter_context(self._types_namespace_stack.push(dataclass_base))
```

Reply:
Please extend the tests so that they fail if the `reversed` call isn't here. I think this requires a child class overriding an existing field in the parent.

Reply:
It took a bit of staring at the code, but the reversed is actually wrong (the merge order `types_namespace_stack.push` uses only adds new names, it doesn't overwrite names already in the stack) so I've removed it (and added the test). Since the phantom modules can't import each other the added test uses multiple inheritance.